## Homework 04
  
The [Conversation AI](https://conversationai.github.io/) team, a research initiative founded by [Jigsaw](https://jigsaw.google.com/) and Google (both a part of Alphabet) are working on tools to help improve online conversation. One area of focus is the study of negative online behaviors, like toxic comments (i.e. comments that are rude, disrespectful or otherwise likely to make someone leave a discussion).   
  
Kaggle are currently hosting their [second competition](https://www.kaggle.com/c/jigsaw-toxic-comment-classification-challenge#description) on this research. The challenge is to create a model that is capable of detecting different types of of toxicity like threats, obscenity, insults, and identity-based hate better than Perspective’s current models. The competitions use a dataset of comments from Wikipedia’s talk page edits. Improvements to the current model will hopefully help online discussion become more productive and respectful.

We shall be using this dataset to benchmark a number of ML models. While the focus of the current competition is to mitigate bias, we will not be using the metric used in the competition. Instead we will be focusing on a simpler metric [Area under the Curve (or AUC)](https://www.kaggle.com/learn-forum/53782) which is suitable to unbalanced binary datasets. Also, we shall not consider different levels of Toxicity; we shall purely take anything marked over the 0.5 level in the measured toxicity range as toxic, and anything underneath as non-toxic. 

We have created a jupyter notbook with some of the tools to model this problem in Deep Learning, using Logistic regression and MLP. Your challenge will be to fill in the models and benchmark the AUC you achieve on these models.

We shall be using the keras deep learning package. As you may know, this is an API into DL frameworks, but is most commonly backed by Tensorflow. [keras.io](keras.io) is a great source for documentation and examples on layers available andn functionality. 

**Have fun!!**


*Disclaimer: the dataset used contains text that may be considered profane, vulgar, or offensive.*

### Set up packages

In [2]:
import os
import tensorflow as tf
import keras
sess = tf.Session(config=tf.ConfigProto(log_device_placement=True))
import pandas as pd
import numpy as np

from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import CountVectorizer 
from sklearn.model_selection import KFold
from sklearn.metrics import roc_auc_score

Using TensorFlow backend.


### Load data
Let's load our data and take a peak.   
The competition metrics and fields have too much detail too cover here.   
We will just focus on the comment of the users and whether it was deemed toxic (target>0.5) or not.    

In [3]:
# Lets look at what files we have available. 
os.listdir()

['.ipynb_checkpoints', 'w251_homework04.ipynb', 'v2', 'train.csv.zip']

In [4]:
train_df = pd.read_csv('train.csv.zip', 
                       dtype={'comment_text':str},
                       usecols=['comment_text', 'target'],
                       compression = 'zip')
train_df['target'] = (train_df['target']>0.5).astype(int)
print("Dataframe Number of (rows, columns) : {}".format(train_df.shape))

Dataframe Number of (rows, columns) : (1804874, 2)


In [4]:
pd.options.display.max_colwidth = 200
train_df.head(10)

target  \
0       0   
1       0   
2       0   
3       0   
4       1   
5       1   
6       0   
7       0   
8       0   
9       0   

                                                                                                               comment_text  
0                     This is so cool. It's like, 'would you want your mother to read this??' Really great idea, well done!  
1        Thank you!! This would make my life a lot less anxiety-inducing. Keep it up, and don't let anyone get in your way!  
2                                    This is such an urgent design problem; kudos to you for taking it on. Very impressive!  
3                                      Is this something I'll be able to install on my site? When will you be releasing it?  
4                                                                                      haha you guys are a bunch of losers.  
5                                                                                                      ur a sh*tty comment.  
6                                                                                               hahahahahahahahhha suck it.  
7                                                                                                       FFFFUUUUUUUUUUUUUUU  
8  The ranchers seem motivated by mostly by greed; no one should have the right to allow their animals destroy public land.  
9                                          It was a great show. Not a combo I'd of expected to be good together but it was.

### Create validation data set
Lets randomly 66/33 split the data into a training and validation set.   
<font color='blue'>**No change needed here - note, please do not change the KFold split parameters, keeping it consistent will help us debug.**</font>

In [5]:
cv = KFold(n_splits=3, shuffle=True, random_state=42)
trn_ids, val_ids = next(cv.split(train_df))
x_train, x_valid = train_df['comment_text'][trn_ids], train_df['comment_text'][val_ids]
y_train, y_valid = train_df['target'].values[trn_ids], train_df['target'].values[val_ids]

### Vectorize Count of tokens

![Count Vectorizer](https://cdn-images-1.medium.com/max/1600/1*LD5XaHzOnoniU4p137FL5g.jpeg)  
We shall start off performing some CPU based Deep Learning operations. Sparse matrices are better run on CPU.   
Do not underestimate CPU based Deep Learning such as MLP; this can be very powerful and outperform complex much more complex DL models.   
Here we create a sparse matrix from the text, with 200K of the most common unigram and bigrams.  
<font color='blue'>
**You task here is to convert the collection of text documents (found in the `comment_text` field) to a matrix of token counts.  
This can be done using the [CountVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html) in scikit_learn.  
After creating the vecotrizer, fit it based on the train matrix `x_train` and use this vectorizer to transform both the `x_train` and `x_valid` sets.   
Create sparse matrices called `X_trn_mat` and `X_val_mat`, and please call your vectorizer: `vectorizer`.  
Use the parameters max features = 200000 and the token pattern `\w+`. This token pattern matches one or more word characters (same as `[a-zA-Z0-9_]`) only. all other characters are stripped.
Also, we would like to count both unigrams and bigrams (pairs of words), so set the ngram range to `(1,2)`**
    </font>

In [6]:
#>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
# Create a CountVectorizer, called `vectorizer`
# And create sparse matrices X_trn_mat & X_val_mat
#<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
vectorizer = CountVectorizer(max_features=200000, token_pattern='\w+',ngram_range=(1,2))

In [7]:
vectorizer.fit(x_train)

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=200000, min_df=1,
                ngram_range=(1, 2), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='\\w+', tokenizer=None,
                vocabulary=None)

In [8]:
X_trn_mat = vectorizer.transform(x_train)

In [9]:
X_val_mat = vectorizer.transform(x_valid)

In [12]:
print(vectorizer.get_feature_names()[:10])
print(vectorizer.get_feature_names()[100000:100000+10])

['0', '0 0', '0 00', '0 01', '0 05', '0 1', '0 2', '0 25', '0 3', '0 4']
['make more', 'make most', 'make much', 'make my', 'make new', 'make no', 'make noise', 'make obama', 'make of', 'make on']


In [13]:
X_trn_mat

<1203249x200000 sparse matrix of type '<class 'numpy.int64'>'
	with 89432534 stored elements in Compressed Sparse Row format>

### Logistic Regression

![Logistic Regression](https://upload.wikimedia.org/wikipedia/commons/6/6d/Exam_pass_logistic_curve.jpeg)
  
Lets start off with a simple Logistic Regression, which is the very basic [sigmoid activation function](https://en.wikipedia.org/wiki/Sigmoid_function) used in DL.  
Notice we have no hidden layers, we take as input the whole sparse matrix, and as output the binary classifier prediction (`0<=output<=1`).  
The model has 200001 parameters. One coefficient per column in the sparse matrx, plus one bias variable - each of which is learned using gradient descent. 


In [14]:
X_trn_mat.shape

(1203249, 200000)

In [37]:
model_in = keras.Input(shape=(X_trn_mat.shape[1],), dtype='float32', sparse=True)
out = keras.layers.Dense(1, activation='sigmoid')(model_in)
model = keras.Model(inputs=model_in, outputs=out)
model.compile(loss='binary_crossentropy', optimizer=keras.optimizers.Adam(lr=1e-2))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_2 (InputLayer)         (None, 200000)            0         
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 200001    
Total params: 200,001
Trainable params: 200,001
Non-trainable params: 0
_________________________________________________________________


In [23]:
model.fit(X_trn_mat, y_train, epochs=2, batch_size=2**13, validation_data=(X_val_mat, y_valid))

Instructions for updating:
Use tf.cast instead.
Train on 1203249 samples, validate on 601625 samples
Epoch 1/2
1203249/1203249 [==============================] - 11s 9us/step - loss: 0.2405 - val_loss: 0.1769
Epoch 2/2
1203249/1203249 [==============================] - 11s 9us/step - loss: 0.1322 - val_loss: 0.1490


In [25]:
preds_lr = model.predict(X_val_mat).flatten()
print('AUC score : {:.5f}'.format(roc_auc_score(y_valid, preds_lr)))

AUC score : 0.88496


Looks at the coefficients to see which words are driving toxic and non-toxic sentences. 

In [26]:
feats = np.array(vectorizer.get_feature_names())
importance_index = model.get_weights()[0].flatten().argsort()
print('Top toxic tokens : \n{}'.format(feats[importance_index[-10:]].tolist()))
print('\nTop non-toxic tokens : \n{}'.format(feats[importance_index[:10]].tolist()))

Top toxic tokens : 
['hypocrites', 'damn', 'shit', 'morons', 'stupid', 'pathetic', 'idiot', 'stupidity', 'idiots', 'idiotic']

Top non-toxic tokens : 
['amen', 'well said', 'my place', 'remember to', 'underscores', 'propaganda from', 'bingo', 'employees as', 'the dsm', 'more conservative']


In [27]:
feats

array(['0', '0 0', '0 00', ..., 'ᴡᴇᴇᴋ http', 'ᴡᴏʀᴋɪɴɢ', 'ᴡᴏʀᴋɪɴɢ ғʀᴏᴍ'],
      dtype='<U52')

### MLP

![MLP](https://www.researchgate.net/profile/Mouhammd_Alkasassbeh/publication/309592737/figure/fig2/AS:423712664100865@1478032379613/MultiLayer-Perceptron-MLP-sturcture-334-MultiLayer-Perceptron-Classifier-MultiLayer.png)

Here we shall create a Multi layer perceptron. Although relatively simple, these can be very powerful models and useful when compute power is low. 
<font color='blue'>**Please add three hidden layers to the network using a `relu` activation function.  
You can refer to this [script](https://www.kaggle.com/lopuhin/mercari-golf-0-3875-cv-in-75-loc-1900-s) which was an MLP taking first place to win the *Mercari Price Suggestion Challenge*. However note, you can do this task by only adding three or four lines. You should see a large increace in performance over the Logistic Regression.**</font>  
Never underestimate the power of an MLP!!

In [18]:
model_in = keras.Input(shape=(X_trn_mat.shape[1],), dtype='float32', sparse=True)
#>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
# Please fill in the next lines with the three hidden layers and the output layer. 
#<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
layer1 = keras.layers.Dense(192, activation='relu')(model_in)
layer2 = keras.layers.Dense(64, activation='relu')(layer1)
layer3 = keras.layers.Dense(64, activation='relu')(layer2)
out = keras.layers.Dense(1, activation='relu')(layer3)
model = keras.Model(inputs=model_in, outputs=out)
model.compile(loss='binary_crossentropy', optimizer=keras.optimizers.Adam(lr=1e-3))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_2 (InputLayer)         (None, 200000)            0         
_________________________________________________________________
dense_5 (Dense)              (None, 192)               38400192  
_________________________________________________________________
dense_6 (Dense)              (None, 64)                12352     
_________________________________________________________________
dense_7 (Dense)              (None, 64)                4160      
_________________________________________________________________
dense_8 (Dense)              (None, 1)                 65        
Total params: 38,416,769
Trainable params: 38,416,769
Non-trainable params: 0
_________________________________________________________________


In [19]:
model.fit(X_trn_mat, y_train, batch_size=2**13, epochs=2, verbose=1, validation_data=(X_val_mat, y_valid))
preds_mlp = model.predict(X_val_mat).flatten()

Train on 1203249 samples, validate on 601625 samples
Epoch 1/2
1203249/1203249 [==============================] - 109s 90us/step - loss: 0.1655 - val_loss: 0.1630
Epoch 2/2
1203249/1203249 [==============================] - 106s 88us/step - loss: 0.0940 - val_loss: 0.1789


In [20]:
print('AUC score : {:.5f}'.format(roc_auc_score(y_valid, preds_mlp)))

AUC score : 0.90537


### MLP with regularisation

Now lets try regularisation.  
<font color='blue'>**Copy the above MLP model and create a new one adding regularisation.  
    Add l2 regularisation to each of the dense layers. Check on [keras.io](keras.io) to find details on how to add l2 regularisation. Play are around with different level of regularisation to see when you achieve optimal results.   
Generally it is good to choose parameters like this by using different factors of `10`.  
Can you improve on you previous AUC results by using reglarisation?**</font>

In [22]:
from keras import regularizers

In [28]:
model_in = keras.Input(shape=(X_trn_mat.shape[1],), dtype='float32', sparse=True)
#>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
# Please fill in the next lines with the three hidden layers and the output layer. 
#<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
layer1 = keras.layers.Dense(192, activation='relu', kernel_regularizer=regularizers.l2(0.001))(model_in)
layer2 = keras.layers.Dense(64, activation='relu', kernel_regularizer=regularizers.l2(0.001))(layer1)
layer3 = keras.layers.Dense(64, activation='relu', kernel_regularizer=regularizers.l2(0.001))(layer2)
out = keras.layers.Dense(1, activation='relu', kernel_regularizer=regularizers.l2(0.001))(layer3)
model = keras.Model(inputs=model_in, outputs=out)
model.compile(loss='binary_crossentropy', optimizer=keras.optimizers.Adam(lr=1e-3))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_6 (InputLayer)         (None, 200000)            0         
_________________________________________________________________
dense_17 (Dense)             (None, 192)               38400192  
_________________________________________________________________
dense_18 (Dense)             (None, 64)                12352     
_________________________________________________________________
dense_19 (Dense)             (None, 64)                4160      
_________________________________________________________________
dense_20 (Dense)             (None, 1)                 65        
Total params: 38,416,769
Trainable params: 38,416,769
Non-trainable params: 0
_________________________________________________________________


In [29]:
model.fit(X_trn_mat, y_train, batch_size=2**13, epochs=2, verbose=1, validation_data=(X_val_mat, y_valid))
preds_mlp_reg = model.predict(X_val_mat).flatten()

Train on 1203249 samples, validate on 601625 samples
Epoch 1/2
1203249/1203249 [==============================] - 121s 100us/step - loss: 0.3935 - val_loss: 0.3413
Epoch 2/2
1203249/1203249 [==============================] - 116s 96us/step - loss: 0.3286 - val_loss: 0.3222


In [30]:
print('AUC score : {:.5f}'.format(roc_auc_score(y_valid, preds_mlp_reg)))

AUC score : 0.92024


A lower value leads to more regularization, so 0.001 works better than 0.01, which works better than 0.1. I was able to get a small increase in AUC using the 0.001 value. Given that our model has ~38M parameters, with only 1203249 training samples, the model is likely to overfit without regularization. I think a dropout layer would do very well also.

In [31]:
model_in = keras.Input(shape=(X_trn_mat.shape[1],), dtype='float32', sparse=True)
#>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
# Please fill in the next lines with the three hidden layers and the output layer. 
#<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
layer1 = keras.layers.Dense(192, activation='relu')(model_in)
layer1 = keras.layers.Dropout(0.3)(layer1)
layer2 = keras.layers.Dense(64, activation='relu')(layer1)
layer2 = keras.layers.Dropout(0.3)(layer2)
layer3 = keras.layers.Dense(64, activation='relu')(layer2)
layer3 = keras.layers.Dropout(0.3)(layer3)
out = keras.layers.Dense(1, activation='relu')(layer3)
model = keras.Model(inputs=model_in, outputs=out)
model.compile(loss='binary_crossentropy', optimizer=keras.optimizers.Adam(lr=1e-3))
model.summary()

Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_7 (InputLayer)         (None, 200000)            0         
_________________________________________________________________
dense_21 (Dense)             (None, 192)               38400192  
_________________________________________________________________
dropout_1 (Dropout)          (None, 192)               0         
_________________________________________________________________
dense_22 (Dense)             (None, 64)                12352     
_________________________________________________________________
dropout_2 (Dropout)          (None, 64)                0         
_________________________________________________________________
dense_23 (Dense)             (None, 64)                4160      
___________________________

In [32]:
model.fit(X_trn_mat, y_train, batch_size=2**13, epochs=2, verbose=1, validation_data=(X_val_mat, y_valid))
preds_mlp_drop = model.predict(X_val_mat).flatten()

Train on 1203249 samples, validate on 601625 samples
Epoch 1/2
1203249/1203249 [==============================] - 109s 91us/step - loss: 0.1945 - val_loss: 0.1424
Epoch 2/2
1203249/1203249 [==============================] - 107s 89us/step - loss: 0.1115 - val_loss: 0.1566


In [33]:
print('AUC score : {:.5f}'.format(roc_auc_score(y_valid, preds_mlp_drop)))

AUC score : 0.89770
